wagon-providers/wagon-http-shared/src/main/java/org/apache/maven/wagon/shared/http/AbstractHttpClientWagon.java

```diff
@@ -169,6 +154,44 @@ private WagonHttpEntity( final InputStream stream, final Resource resource, fina
 
             this.wagon = wagon;
 
+            // if the autoset content flag is SET and the content type is blank
+            // then try to determine what the content type is and set it
+            if ( AUTOSET_CONTENT_TYPE && getContentType() == null )
+            {
+                try
+                {
+                    autosetContentType();
+                }
+                catch ( IOException ioX )
+                {
+                    if ( AUTOSET_CONTENT_TYPE_FATAL )
+                    {
+                        throw new TransferFailedException(
+                                "Failed to determine content type, "
+                                + " unset 'maven.wagon.http.autocontenttype.fatal' to allow continued processing",
+                                ioX );
+                    }
+                }
+            }
+        }
+
+        private void autosetContentType() throws IOException
+        {
+            // if the content-type has not been set then
+            // if the option flag is TRUE and the content type is determinable from the file extension
+            // then set it from the file extension
+            final String mimeType = AUTOSET_CONTENT_TYPE
+                    ? this.source != null
+                        ? URLConnection.guessContentTypeFromName( source.getName() )
+                        : this.stream != null
+                            ? URLConnection.guessContentTypeFromStream( this.stream )
```

Comment:
You cannot do this. This will confuse a few magic bytes from the stream making it useless. You can utilize this only with a pushback stream.

Reply:
If the stream does not support mark then guessContentTypeFromStream will return null. Here is the code:
    static public String guessContentTypeFromStream(InputStream is)
                        throws IOException {
        // If we can't read ahead safely, just give up on guessing
        if (!is.markSupported())
            return null;
...

Reply:
Javadoc does not even mentioned that. Please add a comment above.